docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,287 @@
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses small memory. 
+
+For inactive IndexReaders and BucketFiles, if the inactive duration (`now.Sub(lastQueryTime)`) is greater than a configured duration (e.g. 15 mins), then IndexReaders will remove in-memory objects, and BucketFiles will remove in-memory bitmaps, close file descriptors and unmmap local files. This will reduce memory usage.
+
+### Preload blocks on startup
+
+The latest blocks will be queried the most. 
+
+Preload `the last N hours` (configurable) blocks on startup.
+This will fetch index files from object storage and decode bytes into in-memory objects.
+
+### Local files
+
+Because the max size of a index file is 64 GiB, and the default size of a chunks file is 512 MiB.
+
+So we can set the size of a local index file to 64 GiB, set the size of a local chunks file to 768 (512*1.5) MiB.
+
+The size is used to create a sparse file and mmap a file.
+
+### Object storage byte ranges
+
+Currently, different provider clients implement byte ranges differently. We need to wrap their implementations and export the same api.
+
+ I think byte ranges in [minio-go](https://github.com/minio/minio-go/blob/master/api-get-options.go#L99-L128) is a good example:
+
+1. Fetch last N bytes:     
+     `Range{start:0, end: -N}` ->  Request header`Range: bytes=-N`     
+     **NOTE**: will not write to local file because we do not know the offset.
+2. Fetch everything starting from `start`:    
+    `Range{start:N, end: 0}	`  ->  Request heade `Range: bytes=N-` 
+
+3. Fetch everything starting at `start` till  `end`:    
+   `Range{start:N, end: M}` -> Request header `Range: bytes=M-N` 
+
+Now all providers (Azure, COS, GCS, S3, Swift) support `Fetch last N bytes`, `Fetch everything starting from offset` and `Fetch everything starting at start till end`.
+
+In order to support `Fetch last N bytes` in client side, there are some required changes:
+
+1. Upgrade GCS client package to version [v0.45.0+](https://github.com/googleapis/google-cloud-go/blob/master/CHANGES.md#v0450)
+2. reimplement Swift [GetRange](https://github.com/thanos-io/thanos/blob/master/pkg/objstore/swift/swift.go#L121)
+  
+
+See https://tools.ietf.org/html/rfc7233#section-2.1 for reference.
+
+### Work plan
+
+- A) Reimplement `BucketReader.GetRange` 
+  - Upgrade GCS client package
+  - Reimplement Swift `GetRange` 
+- B) Implement `BucketFile`
+  
+- C) Use `BucketFile` for `BucketIndexReader` and `BucketchunkReader`
+
+- D) Add Flags to enable `BucketFile` 
+
+### Future work
+
+Design a new index file for [postings offset entries](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset entries](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) for sequential and random accesses.
```

Comment:
a new index-cache /index-header I guess? Yes! And it might be the first step /treated separately not future work (:

Reply:
removed.